From ba6c74914827cb5ace2c15a609d8c1be5c553ce8 Mon Sep 17 00:00:00 2001
From: Ricardo Gaviria <ricardo.a.gaviria@gmail.com>
Date: Sat, 4 Aug 2018 22:46:10 +0200
Subject: [PATCH] Handle error when encrypted file is exctracted inside archive

---
 debian/changelog              |  5 +++++
 diffoscope/comparators/zip.py | 10 +++++++---
 2 files changed, 12 insertions(+), 3 deletions(-)

diff --git a/debian/changelog b/debian/changelog
index 93bd3dc..330f6f5 100644
--- a/debian/changelog
+++ b/debian/changelog
@@ -1,3 +1,8 @@
+diffoscope (100.0~reproducible1) UNRELEASED; urgency=medium
+  
+  [Ricardo Gaviria]
+  * Handle error when trying to extract an encrypted file from archive
+
 diffoscope (100) UNRELEASED; urgency=medium
 
   * WIP.
diff --git a/diffoscope/comparators/zip.py b/diffoscope/comparators/zip.py
index e77979c..b033abf 100644
--- a/diffoscope/comparators/zip.py
+++ b/diffoscope/comparators/zip.py
@@ -25,6 +25,7 @@ import zipfile
 
 from diffoscope.tools import tool_required
 from diffoscope.difference import Difference
+from diffoscope.exc import ContainerExtractionError
 
 from .utils.file import File
 from .directory import Directory
@@ -98,9 +99,12 @@ class ZipContainer(Archive):
         # any weird character so we can get to the bytes.
         targetpath = os.path.join(dest_dir, os.path.basename(member_name)).encode(
             sys.getfilesystemencoding(), errors='replace')
-        with self.archive.open(member_name) as source, open(targetpath, 'wb') as target:
-            shutil.copyfileobj(source, target)
-        return targetpath.decode(sys.getfilesystemencoding())
+        try:
+            with self.archive.open(member_name) as source, open(targetpath, 'wb') as target:
+                shutil.copyfileobj(source, target)
+            return targetpath.decode(sys.getfilesystemencoding())
+        except Exception as exc:
+            raise ContainerExtractionError(member_name, exc)
 
     def get_member(self, member_name):
         zipinfo = self.archive.getinfo(member_name)
-- 
2.7.4

